std/range/package.d

```diff
@@ -9594,6 +9594,388 @@ auto refRange(R)(R* range)
     foo(r);
 }
 
+/**
+Bitwise adapter over integral type ranges. Consumes the range elements bit by bit.
+
+Params:
+    R = an input range to iterate over
+
+Returns:
+    At minimum, an input range. If `R` was a forward, bidirectional or random
+    access range, `Bitwise!R` will be as well.
+*/
+struct Bitwise(R)
+    if (isInputRange!R && isIntegral!(ElementType!R))
+{
+private:
+    alias ElemType = ElementType!R;
+    alias UnsignedElemType = Unsigned!ElemType;
+
+    R parent;
+    enum bitsNum = ElemType.sizeof * 8;
+    size_t maskPos = bitsNum;
+
+    static if (isBidirectionalRange!R)
+    {
+        size_t backMaskPos = 1;
+    }
+
+public:
+    this()(auto ref R range)
+    {
+        parent = range;
+    }
+
+    /**
+     * Check if the range is empty
+     *
+     * Returns: a boolean true or false
+     */
+    bool empty()
+    {
+        static if (hasLength!R)
+        {
+            return length() == 0;
+        }
+        else static if (isBidirectionalRange!R)
+        {
+            bool isOverlapping = parent.empty;
+            if (!parent.empty)
+            {
+                /*
+                   If we have consumed the last element of the range both from
+                   the front and the back, then the masks positions will overlap
+                 */
+                R parentCopy = parent.save;
+                parentCopy.popFront;
+                isOverlapping = parentCopy.empty && (maskPos < backMaskPos);
+            }
+            return parent.empty || isOverlapping;
+        }
+        else
+        {
+            /*
+               If we consumed the last element of the range, but not all the
+               bits in the last element
+             */
+            return parent.empty;
+        }
+    }
+
+    bool front()
+    {
+        assert(!empty());
+        return (parent.front & mask(maskPos)) != 0;
+    }
+
+    void popFront()
+    {
+        --maskPos;
+        if (maskPos == 0)
+        {
+            parent.popFront;
+            maskPos = bitsNum;
+        }
+    }
+
+    static if (hasLength!R)
+    {
+        size_t length()
+        {
+            auto len = parent.length * bitsNum - (bitsNum - maskPos);
+            static if (isBidirectionalRange!R)
+            {
+                len -= (backMaskPos - 1);
+            }
+            return len;
+        }
+
+        alias opDollar = length;
+    }
+
+    static if (isForwardRange!R)
+    {
+        typeof(this) save()
+        {
+            auto result = this;
+            result.parent = parent.save;
+            return result;
+        }
+    }
+
+    static if (isBidirectionalRange!R)
+    {
+        bool back()
+        {
+            assert(!empty());
```

Comment:
this function is not covered (please install the Codecov Chrome Extension and you'll see uncovered lines in red on this page)